### DISCUSSION TITLE
ExpressionBased model initial solution

### DISCUSSION THREAD

**Discussion by andreasdahlberg01** [optimatika/ojAlgo#512] [General, ojAlgo and related code]
I am using ojAlgo (version 51.4.1, last one supporting java 8) to solve an integer problem. As far as I know, the `ExpressionBasedModel` can't take an initial solution but `IntegerSolver` can. So I am solving the optimization problem as:

` val result = IntegerSolver.make(expressionModel).solve(initialSolution)
`
I get strange behavior since the `optimisationSense` is `null` but in earlier versions it was `minimize` by default. Also, in this version it is not possible to set the `optimisationSense`.

Any help is appreciated

**Comment by apete** [comment]
When you call `model.minimise()` or `model.maximise()` an initial solution is derived from the model (variables) and supplied to the solver. 

If you do `variable.setValue(...)` that's the value that will be used and you are effectively setting the initial solution.

**Reply by andreasdahlberg01** [reply]
Ah right, thanks!

Another question: what is the preferred way to extract the best solution? I am currently using `result.doubleValue(index)` where `result` is the return from `model.maximise()`, but is it also possible to use `variable.getValue()` after having called `maximise()`?

**Comment by apete** [comment]
Yes, using  `model.minimise()` or `model.maximise()` the solution is written to the variable instances - converted to `BigDecimal` and rounded/enforced using the `options.solution` `NumberContext`.

**Comment by andreasdahlberg01** [comment]
Another unrelated question: In this newer version the solver sometimes prints error messages (I think when generating cuts)

`Obviously infeasible value 167: 0 <= 1.0640204985170896 <= 1 @ 0 (0) -1=NaN NaN ....`

Is this a bug in the solver or is it expected to happen? Is there any way to turn off error messages like that one?

**Comment by apete** [comment]
It's the MIP solver validating the LP node-solution and finding that it failed in a way it shouldn't. The fact that the LP fails that way is a problem, but the MIP usually deals with it. If you get a lot of them you should try to investigate further.

**Reply by andreasdahlberg01** [reply]
I don't get them often but it is annoying to see error messages in the logs. Any pointers on what I can do (settings etc) to prevent this from happening?

**Reply by apete** [reply]
You can't get rid of them without changing the code.
